numpy/core/arrayprint.py

```diff
@@ -468,14 +468,17 @@ def wrapper(self, *args, **kwargs):
 # gracefully handle recursive calls, when object arrays contain themselves
 @_recursive_guard()
 def _array2string(a, options, separator=' ', prefix=""):
-    # The formatter __init__s cannot deal with subclasses yet
-    data = asarray(a)
+    # The formatter __init__s in _get_format_function cannot deal with
+    # subclasses yet, and we also need to avoid recursion issues in
+    # _formatArray with subclasses which return 0d arrays in place of scalars
+    a = asarray(a)
```

Comment:
The change here is the underlying cause for the astropy failure https://github.com/astropy/astropy/issues/7214 -- but reverting it causes the recursion guard to fail.

Reply:
I think the better fix is to move the `asarray` inside the formatters, but that breaks users who decide to pass `str` as a formatter.